sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteSchema.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.impl;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.beam.sdk.extensions.sql.meta.Table;
+import org.apache.beam.sdk.extensions.sql.meta.provider.TableProvider;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.schema.Function;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.SchemaVersion;
+import org.apache.calcite.schema.Schemas;
+
+/**
+ * Adapter from {@link TableProvider} to {@link Schema}.
+ */
+public class BeamCalciteSchema implements Schema {
+  private TableProvider tableProvider;
+
+  public BeamCalciteSchema(TableProvider tableProvider) {
+    this.tableProvider = tableProvider;
+  }
+
+  public TableProvider getTableProvider() {
+    return tableProvider;
+  }
+
+  @Override
+  public boolean isMutable() {
+    return true;
+  }
+
+  @Override
+  public Schema snapshot(SchemaVersion version) {
+    return this;
+  }
+
+  @Override
+  public Expression getExpression(SchemaPlus parentSchema, String name) {
+    return Schemas.subSchemaExpression(parentSchema, name, getClass());
+  }
+
+  @Override
+  public Set<String> getTableNames() {
+    Set<String> tables = new HashSet();
+    for (Table table : tableProvider.listTables()) {
+      tables.add(table.getName());
+    }
+    return tables;
+  }
+
+  @Override
+  public org.apache.calcite.schema.Table getTable(String name) {
+    for (Table table : tableProvider.listTables()) {
+      if (table.getName().equals(name)) {
+        return new BeamCalciteTable(tableProvider.buildBeamSqlTable(table));
+      }
+    }
```

Comment:
nit: looks like it would be better to convert this to a `Map<String, BeamCalciteTable>` once in constructor, this way you wouldn't need to implement `map.keySet()` or `map.get()`

Reply:
I'm assuming this comment is on the line with `tableProvider.listTables()` not `}`? If so, the output of that function can not be cached. I do however agree that it makes sense to change the return value of that API to `Map<String, Table>`. Simplifies a lot of code all over, so I'll do that.